File: DataBaseRoomApp/jroomapp/src/main/java/com/woowrale/jroomapp/usecases/ContactAllUseCase.java
package com.woowrale.jroomapp.usecases;

import com.woowrale.jroomapp.data.repository.LocalRepository;
import com.woowrale.jroomapp.domain.model.Contact;
import com.woowrale.jroomapp.usecases.base.BaseUseCase;
import com.woowrale.jroomapp.usecases.threads.JobScheduler;
import com.woowrale.jroomapp.usecases.threads.UIScheduler;

import java.util.List;

import io.reactivex.Single;

public class ContactAllUseCase extends BaseUseCase<List<Contact>, ContactAllUseCase.Params> {

    private LocalRepository localRepository;

    public ContactAllUseCase(LocalRepository localRepository, UIScheduler uiScheduler, JobScheduler jobScheduler) {
        super(uiScheduler, jobScheduler);
        this.localRepository = localRepository;
    }

    @Override
    protected Single<List<Contact>> buildUseCaseObservable(Params params) {
        return Single.create(emitter -> {
            try {
                saveAllContacstCase(params.getContacts());
                List<Contact> contacts = getAllCase();
                emitter.onSuccess(contacts);
            } catch (Exception exception) {
                if (!emitter.isDisposed()) {
                    emitter.onError(exception);
                }
            }
        });
    }

    public Contact findByIdCase(Integer id) {
        return localRepository.findById(id);
    }

    public List<Contact> getAllCase() {
        return localRepository.getAll();
    }

    public void saveAllContacstCase(List<Contact> contacts) {
        localRepository.saveAll(contacts);
    }

    public void updateContactCase(Contact contact) {
        localRepository.updateContact(contact);
    }

    public static final class Params {

        List<Contact> contacts;

        public Params(List<Contact> contacts) {
            this.contacts = contacts;
        }

        public List<Contact> getContacts() {
            return contacts;
        }
    }
}

File: DataBaseRoomApp/kroomapp/src/main/java/com/woowrale/kroomapp/usecases/ContactAllUseCase.kt
package com.woowrale.kroomapp.usecases

import com.woowrale.kroomapp.data.repository.LocalRepository
import com.woowrale.kroomapp.domain.model.Contact
import com.woowrale.kroomapp.usecases.base.BaseUseCase
import com.woowrale.kroomapp.usecases.threads.JobScheduler
import com.woowrale.kroomapp.usecases.threads.UIScheduler
import io.reactivex.Single

class ContactAllUseCase(private val localRepository: LocalRepository, uiScheduler: UIScheduler, jobScheduler: JobScheduler) : BaseUseCase<List<Contact>, ContactAllUseCase.Params>(uiScheduler, jobScheduler) {

    override fun buildUseCaseObservable(params: Params): Single<List<Contact>> {
        return Single.create {
            try {
                this.saveAllContacstCase(params.contacts)
                val contacts = getAllCase()
                it.onSuccess(contacts)
            } catch (exception: Exception) {
                if (!it.isDisposed) {
                    it.onError(exception)
                }
            }
        }
    }

    fun findByIdCase(id: Int): Contact {
        return localRepository.findById(id)
    }

    fun getAllCase(): List<Contact> {
        return localRepository.getAll()
    }

    fun saveAllContacstCase(contacts: List<Contact>) {
        localRepository.saveAll(contacts)
    }

    fun updateContactCase(contact: Contact) {
        localRepository.updateContact(contact)
    }

    data class Params constructor(val contacts: List<Contact>)
}